### ISSUE TITLE
Codespaces

### ISSUE THREAD

**Issue by josebalius** [google/go-github#2362]
I am planning on implementing support for Codespaces, see: https://docs.github.com/en/rest/codespaces

Using this issue to track that work. A couple of questions:
- Do we prefer one large PR with the whole API implementation or multiple PRs?
- Do we prefer to have multiple issues?

Thanks!

**Comment by gmlewis** [comment]
> * Do we prefer one large PR with the whole API implementation or multiple PRs?

I would prefer a bunch of smaller PRs if it is all the same with you.

> * Do we prefer to have multiple issues?

I'm fine with a single issue.

When we have a lot of things to work on, we often use checkboxes in a single issue like this:

- [ ] item 1
- [ ] item 2

etc... feel free to use them if you find it helpful to keep you organized.  You are welcome to use this issue or open a new one.
You can also go back and edit your first entry if you wish... totally up to you.

Thank you, @josebalius !  We appreciate your contributions!

**Comment by gsmachado** [comment]
@josebalius any progress on this?

**Comment by josebalius** [comment]
@gsmachado I started it but never really got into it much. Are you interested in taking over the issue or in the functionality?

**Comment by gsmachado** [comment]
@josebalius I'm developing something that would need the codespaces API. If I progress with the implementation directly on the go-github repo, I will submit a PR. 😄

**Comment by gsmachado** [comment]
@josebalius do you have any branch with your preliminary code?

**Comment by josebalius** [comment]
I have some code locally that I haven't pushed up. Are you interested in it?

**Comment by artificial-aidan** [comment]
Can you push up whatever you came up with?

**Comment by artificial-aidan** [comment]
I started with some simple implementation of codespace management. I'll post the punchlist below. Happy to split it into a new issue as well.

**Comment by artificial-aidan** [comment]
- [ ] Codespaces
  - [ ] List codespaces in a repository for the authenticated user
  - [ ] Create a codespace in a repository
  - [ ] List devcontainer configurations in a repository for the authenticated user
  - [ ] Get default attributes for a codespace
  - [ ] Create a codespace from a pull request
  - [ ] List codespaces for the authenticated user
  - [ ] Create a codespace for the authenticated user
  - [ ] Get a codespace for the authenticated user
  - [ ] Update a codespace for the authenticated user
  - [ ] Delete a codespace for the authenticated user
  - [ ] Export a codespace for the authenticated user
  - [ ] Get dtails about a codespace export
  - [ ] Create a repository from an unpublished codespace
  - [ ] Start a codespace for the authenticated user
  - [ ] Stop a codespace for the authenticated user

**Comment by artificial-aidan** [comment]
@gmlewis I have another branch for secret management, but still waiting on getting the CLA signed. Should I push that commit to the open PR I have, or just wait until I can get the first one merged.

**Comment by gmlewis** [comment]
Your call... whichever is most convenient for you.  I'm fine either way.

**Comment by artificial-aidan** [comment]
The impact area was pretty small. I'll add them to the existing PR while waiting for CLA followup

**Comment by KenSpur** [comment]
Nice, also looking forward to this one. Its needed for the terraform provider

**Comment by artificial-aidan** [comment]
I got the CLA approved, just working through getting me added to it.

**Comment by Not-Dhananjay-Mishra** [comment]
@gmlewis is this issue available? I'd like to add support for the remaining endpoints

**Comment by gmlewis** [comment]
> [@gmlewis](https://github.com/gmlewis) is this issue available? I'd like to add support for the remaining endpoints

Thanks, @Not-Dhananjay-Mishra! I think we can safely assign it to you.

**Comment by Not-Dhananjay-Mishra** [comment]
@gmlewis All remaining endpoints have been added, so this issue can be closed.
